rfcs/react-components/styles-handbook.md

```diff
@@ -0,0 +1,771 @@
+
+```js
+import { makeResetStyles } from '@griffel/react';
+
+const useBaseClassname = makeResetStyles({
+  padding: '4px',
+  margin: '4px',
+});
+```
+
+> ⚠️ **Note:** Only one class generated by `makeResetStyles()` can be applied to an element. Otherwise, behavior will be non-deterministic as classes merging will not be done for this case and results depend on order of insertion.
+
+### Hybrid approach
+
+Our recommendation is use [`makeResetStyles` API][griffel-make-reset-styles] to define base styles for a component and keep using `makeStyles` for overrides and permutations:
+
+```jsx
+import { makeStyles, makeResetStyles, mergeClasses, shorthands } from '@griffel/react';
+
+const useBaseClassname = makeResetStyles({
+  ':hover': {
+    ...shorthands.padding('4px'),
+    /* other styles */
+  },
+  '::before': {
+    display: 'block',
+    content: "' '",
+  },
+  '@media (forced-colors: active)': {
+    ...shorthands.padding('4px'),
+    /* other styles */
+  },
+});
+// ⬇️⬇️⬇️
+// produces 1 class
+
+const useClasses = makeStyles({
+  circular: {
+    ...shorthands.borderRadius('10px'),
+  },
+  primary: {
+    color: 'pink',
+  },
+});
+// ⬇️⬇️⬇️
+// produces 4/1 classes to be conditionally applied
+
+function Component(props) {
+  const baseClassName = useBaseClassname();
+  const classes = useClasses();
+
+  const className = mergeClasses(baseClassName, props.circular && classes.circular, props.primary && classes.primary);
+
+  return <div className={className} />;
+}
+```
+
+# Advanced
+
+## Understanding selectors complexity
+
+CSS Selectors are matched by browser engines from [right to left (bottom-up parsing)][stackoverflow-selectors-match]:
+
+```mermaid
+flowchart RL
+    C[li] --> B[ul] --> A[.menu]
+```
+
+```css
+.menu ul li {
+  color: #00f;
+}
+```
+
+> The browser first checks for `li`, then `ul`, and then `.menu`.
+
+It means that we need to understand selectors that we are writing and avoid _wide_ selectors. A component below will be used as an example:
+
+```jsx
+import { makeStyles } from '@griffel/react';
+
+const useClasses = makeStyles({
+  test: {
+    color: 'orange',
+
+    '> *': {
+      color: 'red',
+    },
+    '> h1': {
+      color: 'magenta',
+    },
+    '> div': {
+      color: 'green',
+    },
+    '> .ui-button': {
+      color: 'blue',
+    },
+  },
+});
+
+function App() {
+  const classes = useClasses();
+
+  return (
+    <>
+      <div className={classes.test}>
+        <h1>Hello World</h1>
+        <button class="ui-button">A button</button>
+      </div>
+
+      {/* Renders 500 empty divs */}
+      {Array.from({ length: 500 }, (_, i) => (
+        <div key={i} />
+      ))}
+    </>
+  );
+}
+```
+
+#### Measurements
+
+##### No selector (best)
+
+Matches only an element where is applied.
+
+- selector `.fe3e8s9`
+- match_attempts **1**
+- match_count **1**
+
+##### `> *` (worst)
+
+"\*" matches any element on a page.
+
+- selector `.fzbuleu > *`
+- match_attempts **502**
+- match_count **1**
```

Comment:
The math does not work here for me. There is no `div` inside the `.fzbuleu`, so what is matching? I guess you want to use `div` or `button` both in the `.fzbuleu` and the loop.

Reply:
I updated stats in 87a9eaa, now it should make sense.

![image](https://user-images.githubusercontent.com/14183168/212699716-92455cb2-5cb2-4ac2-a6ad-6a0874235ab4.png)

It matches inner `button` and `h1`:

![image](https://user-images.githubusercontent.com/14183168/212700071-7d8fcd8f-e66f-41c6-972f-3ec8af99b5a2.png)


If you want to play with it: https://codesandbox.io/s/thirsty-haslett-kqfj87